readthedocs/doc_builder/environments.py

```diff
@@ -207,105 +76,471 @@ def run(self, cmd_input=None, combine_output=False):
         '''
         log.info("Running: '%s' [%s]", self.get_command(), self.cwd)
 
-        stdin = None
         stdout = subprocess.PIPE
         stderr = subprocess.PIPE
-        if cmd_input is not None:
+        stdin = None
+        if self.input_data is not None:
             stdin = subprocess.PIPE
-        if combine_output:
+        if self.combine_output:
             stderr = subprocess.STDOUT
-        proc = subprocess.Popen(
-            self.get_command(),
-            shell=self.shell,
-            cwd=self.cwd,
-            stdin=stdin,
-            stdout=stdout,
-            stderr=stderr,
-            env=self.environment,
-        )
-        cmd_output = proc.communicate(cmd_input)
-        (self.output, self.error) = cmd_output
-        self.status = proc.returncode
+
+        environment = {}
+        environment.update(self.environment)
+        environment['READTHEDOCS'] = 'True'
+        if 'DJANGO_SETTINGS_MODULE' in environment:
+            del environment['DJANGO_SETTINGS_MODULE']
+        if 'PYTHONPATH' in environment:
+            del environment['PYTHONPATH']
+
+        try:
+            proc = subprocess.Popen(
+                self.command,
+                shell=self.shell,
+                cwd=self.cwd,
+                stdin=stdin,
+                stdout=stdout,
+                stderr=stderr,
+                env=environment,
+            )
+            cmd_input = None
+            if self.input_data is not None:
+                cmd_input = self.input_data
+
+            cmd_output = proc.communicate(input=cmd_input)
+            (cmd_stdout, cmd_stderr) = cmd_output
+            try:
+                self.output = cmd_stdout.decode('utf-8', 'replace')
+            except (TypeError, AttributeError):
+                self.output = None
+            try:
+                self.error = cmd_stderr.decode('utf-8', 'replace')
+            except (TypeError, AttributeError):
+                self.error = None
+            self.status = proc.returncode
+        except OSError:
+            self.error = traceback.format_exc()
+            self.output = self.error
+            self.status = -1
 
     def get_command(self):
         '''Flatten command'''
-        if isinstance(self.command, list):
+        if hasattr(self.command, '__iter__'):
```

Comment:
This works in Python 2, as strings have no `__iter__` method. In Python 3, it will return `True` and therefore will insert a space between each char of the command. One solution to that is to add `not isinstance(self.command, str)`.